File: oop-fundamentals/enum/cpp/enum.cpp
#include <iostream>
#include <string>

// Ref - https://www.geeksforgeeks.org/cpp/enum-classes-in-c-and-their-advantage-over-enum-datatype/

//
// =======================================================
// 1. UNscoped ENUM (Legacy C++ Style)
// =======================================================
//

// Old-style enum (avoid in modern C++)
enum Color {
    Red,
    Green,
    Blue
};

//
// =======================================================
// 2. SCOPED ENUM (enum class — Modern C++)
// =======================================================
//

// Strongly typed enum (recommended)
enum class Coin {
    Penny,
    Nickel,
    Dime,
    Quarter
};

//
// =======================================================
// 3. ENUM + BEHAVIOR (Helper Functions)
// =======================================================
//

// Attach meaning via free functions
constexpr int coinValue(Coin coin) {
    switch (coin) {
        case Coin::Penny:   return 1;
        case Coin::Nickel:  return 5;
        case Coin::Dime:    return 10;
        case Coin::Quarter: return 25;
    }
    return 0; // defensive
}

std::string toString(Coin coin) {
    switch (coin) {
        case Coin::Penny:   return "Penny";
        case Coin::Nickel:  return "Nickel";
        case Coin::Dime:    return "Dime";
        case Coin::Quarter: return "Quarter";
    }
    return "Unknown";
}

//
// =======================================================
// 4. ENUM USED IN OBJECT-ORIENTED CODE
// =======================================================
//

class Wallet {
public:
    void addCoin(Coin coin) {
        total_ += coinValue(coin);
    }

    int total() const {
        return total_;
    }

private:
    int total_ = 0;
};

//
// =======================================================
// 5. JAVA-STYLE SINGLETON ENUM EMULATION (NOT RECOMMENDED)
// =======================================================
//

class CoinObject {
public:
    static const CoinObject Penny;
    static const CoinObject Nickel;
    static const CoinObject Dime;
    static const CoinObject Quarter;

    int value() const { return value_; }

private:
    const int value_;

    // Private constructor prevents new instances
    constexpr explicit CoinObject(int v) : value_(v) {}
};

// Static object definitions
constexpr CoinObject CoinObject::Penny{1};
constexpr CoinObject CoinObject::Nickel{5};
constexpr CoinObject CoinObject::Dime{10};
constexpr CoinObject CoinObject::Quarter{25};

//
// =======================================================
// 6. DEMONSTRATION
// =======================================================
//

int main() {

    // ---- Legacy enum problems ----
    Color c = Red;
    int x = c;  // implicit conversion (dangerous)
    std::cout << "Legacy enum converted to int: " << x << "\n";

    // ---- enum class usage ----
    Coin coin1 = Coin::Dime;
    Coin coin2 = Coin::Quarter;

    std::cout << toString(coin1)
              << " value = "
              << coinValue(coin1)
              << "\n";

    // ---- enum used in OOP ----
    Wallet wallet;
    wallet.addCoin(coin1);
    wallet.addCoin(coin2);

    std::cout << "Wallet total (cents): "
              << wallet.total()
              << "\n";

    // ---- Java-style singleton object usage ----
    int objTotal = CoinObject::Dime.value()
                 + CoinObject::Quarter.value();

    std::cout << "Object-based coin total: "
              << objTotal
              << "\n";

    return 0;
}

File: 1.oop-fundamentals/2.enum/ts/enum.ts
//
// =======================================================
// 1. NUMERIC ENUM (BASIC / LEGACY STYLE)
// =======================================================
//

enum Color {
    Red,
    Green,
    Blue
}

// Problems:
// - Treated as numbers
// - Allows meaningless arithmetic
// - Reverse mapping exists at runtime

//
// =======================================================
// 2. NUMERIC ENUM WITH EXPLICIT VALUES
// =======================================================
//

enum CoinEnum {
    Penny = 1,
    Nickel = 5,
    Dime = 10,
    Quarter = 25
}

// Simple usage
const enumTotal = CoinEnum.Dime + CoinEnum.Quarter;

//
// =======================================================
// 3. const enum (COMPILE-TIME ONLY ENUM)
// =======================================================
//

const enum CoinConstEnum {
    Penny = 1,
    Nickel = 5,
    Dime = 10,
    Quarter = 25
}

// Inlined at compile time
const constEnumTotal =
    CoinConstEnum.Dime + CoinConstEnum.Quarter;

//
// =======================================================
// 4. ENUM + BEHAVIOR (LOOKUP FUNCTION)
// =======================================================
//

function coinValue(coin: CoinEnum): number {
    return coin;
}

//
// =======================================================
// 5. ENUM USED IN OOP CONTEXT
// =======================================================
//

class Wallet {
    private total = 0;

    addCoin(coin: CoinEnum) {
        this.total += coinValue(coin);
    }

    getTotal(): number {
        return this.total;
    }
}

const wallet = new Wallet();
wallet.addCoin(CoinEnum.Dime);
wallet.addCoin(CoinEnum.Quarter);

//
// =======================================================
// 6. JAVA-STYLE SINGLETON ENUM EMULATION (CLASS-BASED)
// =======================================================
//

class CoinObject {
    static readonly Penny = new CoinObject(1);
    static readonly Nickel = new CoinObject(5);
    static readonly Dime = new CoinObject(10);
    static readonly Quarter = new CoinObject(25);

    private constructor(private readonly value: number) {}

    getValue(): number {
        return this.value;
    }
}

const objectTotal =
    CoinObject.Dime.getValue() +
    CoinObject.Quarter.getValue();

//
// =======================================================
// 7. MODERN TYPESCRIPT ALTERNATIVE (as const OBJECT)
// =======================================================
//

const Coin = {
    Penny: { value: 1 },
    Nickel: { value: 5 },
    Dime: { value: 10 },
    Quarter: { value: 25 }
} as const;

type Coin = typeof Coin[keyof typeof Coin];

const modernTotal =
    Coin.Dime.value + Coin.Quarter.value;

// ---- Outputs ----
console.log(enumTotal);
console.log(constEnumTotal);
console.log(wallet.getTotal());
console.log(objectTotal);
console.log(modernTotal);
